src/core/internal/utf.d

```diff
@@ -425,61 +425,98 @@ dchar decode(in dchar[] s, ref size_t idx)
+/**
+ * Encodes `c` into the static array `buf`.
+ *
+ * Params:
+ *     buf = destination of encoded character
+ *     c = character to encode
+ *
+ * Returns:
+ *     The length of the encoded character (a number between `1` and `4` for
+ *     `char[4]` buffers and a number between `1` and `2` for `wchar[2]` buffers)
+ *     or `0` in case of failure.
  */
-@safe pure nothrow
-void encode(ref char[] s, dchar c)
-    in
+@nogc nothrow pure @safe
+size_t encode(out char[4] buf, dchar c)
+in
+{
+    assert(isValidDchar(c));
+}
+do
+{
+    if (c <= 0x7F)
     {
-        assert(isValidDchar(c));
+        buf[0] = cast(char) c;
+        return 1;
     }
-    do
+    else if (c <= 0x7FF)
     {
-        char[] r = s;
-
-        if (c <= 0x7F)
-        {
-            r ~= cast(char) c;
-        }
-        else
-        {
-            char[4] buf;
-            uint L;
+        buf[0] = cast(char)(0xC0 | (c >> 6));
+        buf[1] = cast(char)(0x80 | (c & 0x3F));
+        return 2;
+    }
+    else if (c <= 0xFFFF)
+    {
+        buf[0] = cast(char)(0xE0 | (c >> 12));
+        buf[1] = cast(char)(0x80 | ((c >> 6) & 0x3F));
+        buf[2] = cast(char)(0x80 | (c & 0x3F));
+        return 3;
+    }
+    else if (c <= 0x10FFFF)
+    {
+        buf[0] = cast(char)(0xF0 | (c >> 18));
+        buf[1] = cast(char)(0x80 | ((c >> 12) & 0x3F));
+        buf[2] = cast(char)(0x80 | ((c >> 6) & 0x3F));
+        buf[3] = cast(char)(0x80 | (c & 0x3F));
+        return 4;
+    }
+    return 0;
+}
 
-            if (c <= 0x7FF)
-            {
-                buf[0] = cast(char)(0xC0 | (c >> 6));
-                buf[1] = cast(char)(0x80 | (c & 0x3F));
-                L = 2;
-            }
-            else if (c <= 0xFFFF)
-            {
-                buf[0] = cast(char)(0xE0 | (c >> 12));
-                buf[1] = cast(char)(0x80 | ((c >> 6) & 0x3F));
-                buf[2] = cast(char)(0x80 | (c & 0x3F));
-                L = 3;
-            }
-            else if (c <= 0x10FFFF)
-            {
-                buf[0] = cast(char)(0xF0 | (c >> 18));
-                buf[1] = cast(char)(0x80 | ((c >> 12) & 0x3F));
-                buf[2] = cast(char)(0x80 | ((c >> 6) & 0x3F));
-                buf[3] = cast(char)(0x80 | (c & 0x3F));
-                L = 4;
-            }
-            else
-            {
-                assert(0);
-            }
-            r ~= buf[0 .. L];
-        }
-        s = r;
+/// ditto
+@nogc nothrow pure @safe
+size_t encode(out wchar[2] buf, dchar c)
+in
+{
+    assert(isValidDchar(c));
+}
+do
+{
+    if (c <= 0xFFFF)
+    {
+        buf[0] = cast(wchar) c;
+        return 1;
     }
+    else if (c <= 0x10FFFF)
+    {
+        buf[0] = cast(wchar) ((((c - 0x10000) >> 10) & 0x3FF) + 0xD800);
+        buf[1] = cast(wchar) (((c - 0x10000) & 0x3FF) + 0xDC00);
+        return 2;
+    }
+    return 0;
+}
 
+/**
+ * Encodes character c and appends it to array s[].
+ */
+nothrow pure @safe
+void encode(ref char[] s, dchar c)
```

Comment:
Not a fan of the following functions because:
1. They are an overload of the previous `encode` functions, but do something fundamentally different
2. They rely on the GC. We should be crafting functions that do not allocate, but instead send the characters to a sink or an output range.

Reply:
Exactly: https://github.com/dlang/druntime/pull/2404#discussion_r244679517.

Reply:
Also, is this the right order of arguments for UFCS? Would it be better to pass `c` first?

Reply:
The sink normally is the first argument, as in:
```
sink.put(c);
```